{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Edit Distance, Assembly, and Overlaps\n",
    "\n",
    "This notebook explores the theory and application of edit distance and sequence overlaps to answer the following questions (my answers are also provided):\n",
    "\n",
    "- Question 1: What is the edit distance of the best match between pattern `GCTGATCGATCGTACG` and the excerpt of human chromosome 1?  (Don't consider reverse complements.)\n",
    "    - 3\n",
    "\n",
    "- Question 2: What is the edit distance of the best match between pattern \n",
    "`GATTTACCAGATTGAG` and the excerpt of human chromosome 1?  (Don't consider reverse complements.)\n",
    "    - 2\n",
    "\n",
    "- Question 3: Create an overlap graph between the provided fastq file and the exerpt of human chromosome 1. How many edges are in the graph? In other words, how many distinct pairs of reads overlap?\n",
    "    - 904745\n",
    "\n",
    "- Question 4: Picture the overlap graph corresponding to the overlaps computed for the previous question. How many nodes in this graph have at least one outgoing edge? In other words, how many reads have a suffix involved in an overlap?\n",
    "    - 7161\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read genome from FASTA file in folder (chr1.GRCh38.excerpt.fasta) and prepare pattern matching variables\n",
    "\n",
    "# Set FASTA filename\n",
    "filename = 'chr1.GRCh38.excerpt.fasta'\n",
    "\n",
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "\n",
    "# Set text 't' to the genome from file\n",
    "t = readGenome(filename)\n",
    "\n",
    "# Specify shorter text 't' for testing\n",
    "# t = 'TATTGGCTATACGGTT'\n",
    "\n",
    "# Set pattern 'p' to desired string to match\n",
    "p = 'GCTGATCGATCGTACG'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read sequences and qualities from imported FASTQ file\n",
    "\n",
    "# Set FASTQ filename\n",
    "filename = 'ERR266411_1.for_asm.fastq'\n",
    "\n",
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename, 'r') as fq:\n",
    "        while True:\n",
    "            fq.readline()  # skip name line\n",
    "            seq = fq.readline().rstrip()  # read base sequence\n",
    "            fq.readline()  # skip placeholder line\n",
    "            qual = fq.readline().rstrip() # base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities\n",
    "\n",
    "sequences, qualities = readFastq(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n"
     ]
    }
   ],
   "source": [
    "pattern_len = len(p)\n",
    "min_dist = None\n",
    "\n",
    "def editDistance(x, y):\n",
    "    # Create distance matrix\n",
    "    D = []\n",
    "    for i in range(len(x)+1):\n",
    "        D.append([0]*(len(y)+1))\n",
    "    # Initialize first row and column of matrix\n",
    "    for i in range(len(x)+1):\n",
    "        D[i][0] = i\n",
    "    for i in range(len(y)+1):\n",
    "        D[0][i] = i\n",
    "    # Fill in the rest of the matrix\n",
    "    for i in range(1, len(x)+1):\n",
    "        for j in range(1, len(y)+1):\n",
    "            distHor = D[i][j-1] + 1\n",
    "            distVer = D[i-1][j] + 1\n",
    "            if x[i-1] == y[j-1]:\n",
    "                distDiag = D[i-1][j-1]\n",
    "            else:\n",
    "                distDiag = D[i-1][j-1] + 1\n",
    "            D[i][j] = min(distHor, distVer, distDiag)\n",
    "    # Edit distance is the value in the bottom right corner of the matrix\n",
    "    return D[-1][-1]\n",
    "\n",
    "for i in range(len(t) - pattern_len + 1):\n",
    "    substring = t[i:i+pattern_len]\n",
    "    dist = editDistance(p, substring)\n",
    "    if (min_dist is None) or (dist < min_dist):\n",
    "        min_dist = dist\n",
    "\n",
    "print(min_dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total reads loaded: 10000\n",
      "Total overlaps found: 904746\n",
      "Reads with at least one outgoing edge: 7161\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "import urllib.request\n",
    "\n",
    "# Download FASTQ file\n",
    "url = \"https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/ERR266411_1.for_asm.fastq\"\n",
    "filename = \"ERR266411_1.for_asm.fastq\"\n",
    "urllib.request.urlretrieve(url, filename)\n",
    "\n",
    "def read_fastq(fname):\n",
    "    sequences = []\n",
    "    with open(fname, 'r') as fh:\n",
    "        while True:\n",
    "            fh.readline()  # skip header\n",
    "            seq = fh.readline().strip()\n",
    "            fh.readline()  # skip +\n",
    "            fh.readline()  # skip quality\n",
    "            if not seq:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "    return sequences\n",
    "\n",
    "reads = read_fastq(filename)\n",
    "print(\"Total reads loaded:\", len(reads))\n",
    "\n",
    "k = 30\n",
    "\n",
    "# Build prefix index: map prefix (first k bases) -> set of read indices\n",
    "prefix_index = defaultdict(set)\n",
    "for i, read in enumerate(reads):\n",
    "    prefix_index[read[:k]].add(i)\n",
    "\n",
    "def overlap(a, b, min_length):\n",
    "    \"\"\"Return length of longest suffix of a matching prefix of b (â‰¥ min_length)\"\"\"\n",
    "    max_olen = 0\n",
    "    start = 0\n",
    "    while True:\n",
    "        start = a.find(b[:min_length], start)\n",
    "        if start == -1:\n",
    "            break\n",
    "        olen = len(a) - start\n",
    "        if olen >= min_length and b.startswith(a[start:]):\n",
    "            if olen > max_olen:\n",
    "                max_olen = olen\n",
    "        start += 1\n",
    "    return max_olen\n",
    "\n",
    "seen = {}\n",
    "outgoing = set()\n",
    "\n",
    "for i, read in enumerate(reads):\n",
    "    read_len = len(read)\n",
    "    # Try all suffixes length >= k\n",
    "    for start in range(read_len - k, -1, -1):\n",
    "        suffix = read[start:]\n",
    "        if len(suffix) < k:\n",
    "            continue\n",
    "        prefix = suffix[:k]\n",
    "        candidates = prefix_index.get(prefix, set())\n",
    "        for j in candidates:\n",
    "            if i == j:\n",
    "                continue\n",
    "            olen = overlap(read, reads[j], min_length=k)\n",
    "            if olen >= k:\n",
    "                # Update with longest overlap found\n",
    "                if (i, j) not in seen or olen > seen[(i, j)]:\n",
    "                    seen[(i, j)] = olen\n",
    "                    outgoing.add(i)\n",
    "\n",
    "print(\"Total overlaps found:\", len(seen))\n",
    "print(\"Reads with at least one outgoing edge:\", len(outgoing))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
