# 1. Basic Unix Commands

## File Setup

### Direct to project1 folder

\[guest\@centos6\] cd project1

### Extract files

\[guest\@centos6 project1\]\$ gunzip gencommand_proj1_data.tar.gz

\[guest\@centos6 project1\]\$ tar -xf gencommand_proj1_data.tar

### Direct to gencommand_proj1_data folder

\[guest\@centos6 project1\]\$ cd gencommand_proj1_data

\[guest\@centos6 gencommand_proj1_data\]\$ ls

## Questions

### Question 1: How many chromosomes are there in the genome?

\[guest\@centos6 gencommand_proj1_data\]\$ grep -c \"\>\" apple.genome

### Question 2: How many genes and transcript variants?

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f1 apple.genes \| uniq
\| wc -l (or cut -f1 apple.genes \| sort -u \| wc -l)

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f2 apple.genes \| uniq
\| wc -l (or cut -f2 apple.genes \| sort -u \| wc -l)

### Question 3: How many genes have a single splice variant?

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f1 apple.genes \| uniq
-c \| grep " 1 " \| wc -l

### Question 4: How may genes have 2 or more splice variants?

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f1 apple.genes \| uniq
-c \| grep -v " 1 " \| wc -l

### Question 5: How many genes are there on the '+' and '-' strands,
respectively?

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f1,4 apple.genes \| grep
\"+\" \| uniq \| wc -l

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f1,4 apple.genes \| grep
\"-\" \| uniq \| wc -l

### Question 6: How many genes are there on each chromosome? List them in the
order in which they are listed in the genome, i.e. genes on chr1, chr2,
etc.

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f1,3 apple.genes \| grep
\"chr1\" \| uniq \| wc -l

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f1,3 apple.genes \| grep
\"chr2" \| uniq \| wc -l

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f1,3 apple.genes \| grep
\"chr3" \| uniq \| wc -l

### Question 7: How many transcripts are there on each chromosome? Follow
instructions above.

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f2,3 apple.genes \| grep
\"chr1\" \| uniq \| wc -l

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f2,3 apple.genes \| grep
\"chr2" \| uniq \| wc -l

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f2,3 apple.genes \| grep
\"chr3" \| uniq \| wc -l

### Question 8: How many genes are in common between condition A and
condition B?

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f1 apple.conditionA \|
sort -u \> conditionA

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f1 apple.conditionB \|
sort -u \> conditionB

\[guest\@centos6 gencommand_proj1_data\]\$ comm -1 -2 conditionA
conditionB \| wc -l

### Question 9: How many genes are specific to condition A? How many genes
are specific to condition B?

\[guest\@centos6 gencommand_proj1_data\]\$ comm -2 -3 conditionA
conditionB \| wc -l

\[guest\@centos6 gencommand_proj1_data\]\$ comm -1 -3 conditionA
conditionB \| wc -l

### Question 10: How many genes are in common to all three conditions?

\[guest\@centos6 gencommand_proj1_data\]\$ comm -1 -2 conditionA
conditionB \> conditionAB

\[guest\@centos6 gencommand_proj1_data\]\$ cut -f1 apple.conditionC \|
sort -u \> conditionC

\[guest\@centos6 gencommand_proj1_data\]\$ comm -1 -2 conditionAB
conditionC \| wc -l
