# Examine Epigenomics Roadmap data on the human H1 stem cell line; the Roadmap id number is “E003”. 
- We will use data on build “hg19” of the human genome, and for peaks we will use the “narrowPeak” quantification. 
- To make the code easier to run, in several questions we will restrict our analysis to chromosome 22 so as to not exhaust our computational resources. 
- Otherwise results are to be reported on the standard human autosomes (chromosomes 1 to 22).

## Load Required Libraries

```{r}
library(AnnotationHub)
library(Biostrings)
library(BSgenome)
library(GenomicRanges)
library(GenomicFeatures)
library(rtracklayer)
library(Rsamtools)
```

## Question 1: What is the GC content of “chr22” in the “hg19” build of the human genome?
*Tip: The reference genome includes “N” bases; you will need to exclude those.*

```{r}
# load hg19 build dataset
library(BSgenome.Hsapiens.UCSC.hg19)
Hsapiens
# count total bases on chr22
totalBase <- letterFrequency(Hsapiens$chr22, "A") +
             letterFrequency(Hsapiens$chr22, "C") +
             letterFrequency(Hsapiens$chr22, "G") +
             letterFrequency(Hsapiens$chr22, "T")
# count GC bases on chr22  
gcBase <- letterFrequency(Hsapiens$chr22, "GC")
#calculate GC ratio on chr22
gcContent <- gcBase/totalBase
gcContent
```

- 0.4798807

## Question 2:What is mean GC content of H3K27me3 “narrowPeak” regions from Epigenomics Roadmap from the H1 stem cell line on chr 22.
*Background: In the previous assessment we studied H3K27me3 “narrowPeak” regions from the H1 cell line (recall that the Roadmap ID for this cell line is “E003”).* 
*We want to examine whether the GC content of the regions influence the signal; in other words wether the reported results appear biased by GC content.*

```{r}
# find H3K27me3 dataset
ah <- AnnotationHub()
H3K27me3_qh <- query(ah, c("H3K27me3", "E003", "narrowPeak"))
H3K27me3_qh
# retrieve data
H3K27me3_data <- H3K27me3_qh[["AH29892"]]
# get genomic ranges on chr22
H3K27me3_data.chr22 <- subset(H3K27me3_data, seqnames == "chr22")
# get DNA seqence on chr22
H3K27me3_data.chr22.seq <- Views(Hsapiens, H3K27me3_data.chr22)
# calculate GC content on chr22
gcContents <- letterFrequency(H3K27me3_data.chr22.seq, "GC", as.prob = T)
meanGcContents <- mean(gcContents)
meanGcContents
```

- 0.528866

## Question 3: What is the correlation between GC content and “signalValue” of these regions (on chr22)?
*The “narrowPeak” regions includes information on a value they call “signalValue”.*

```{r}
# values of signalValues
sigV <- mcols(H3K27me3_data.chr22.seq)$signalValue
# calculate correlation
cor(sigV, gcContents)
```

- 0.004467924

## Question 4: what is the correlation between the “signalValue” of the “narrowPeak” regions and the average “fc.signal” across the same regions?
*The “narrowPeak” regions are presumably reflective of a ChIP signal in these regions.* 
*To confirm this, we want to obtain the “fc.signal” data from AnnotationHub package on the same cell line and histone modification.* 
*This data represents a vector of fold-change enrichment of ChIP signal over input.*

```{r}
# find fc.signal file
H3K27me3_fc <- query(ah, c("H3K27me3", "E003", "fc.signal"))
H3K27me3_fc
# retrieve data
H3K27me3_fc.data <- H3K27me3_fc[["AH32033"]]
# get subset data on chr22
gr.chr22 <- GRanges(seqnames = "chr22", ranges = IRanges(start =1, end = 51304566))
H3K27me3_fc.rel <- import(H3K27me3_fc.data, which =gr.chr22, as = "Rle")
H3K27me3_fc.rel.chr22 <- H3K27me3_fc.rel$chr22
# view fc.signal data
fc.signal.chr22 <- Views(H3K27me3_fc.rel.chr22, start = start(H3K27me3_data.chr22), end = end(H3K27me3_data.chr22))
# calculate mean values
fc.signal.mean <- mean(fc.signal.chr22)
# calculate correlation
cor(fc.signal.mean, sigV)
```

- 0.9149614

## Question 5: Question: How many bases on chr22 have an fc.signal greater than or equal to 1?

```{r}
sum(H3K27me3_fc.rel.chr22 >= 1)
```

- 10914671

## Question 6: Identify the regions of the genome where the signal in E003 is 0.5 or lower and the signal in E055 is 2 or higher.
*The H1 stem cell line is an embryonic stem cell line, a so-called pluripotent cell. Many epigenetic marks change upon differentiation.* 
*We will examine this. We choose the cell type with Roadmap ID “E055” which is foreskin fibroblast primary cells.*
*We will use the “fc.signal” for this cell type for the H3K27me3 mark, on chr22. We now have a signal track for E003 and a signal track for E055.* 
*We want to identify regions of the genome which gain H3K27me3 upon differentiation.* 
*These are regions which have a higher signal in E055 than in E003. To do this properly, we would need to standardize (normalize) the signal across the two samples; we will ignore this for now.*

```{r}
H3K27me3_fc_E055 <- query(ah, c("H3K27me3", "E055"))
H3K27me3_fc_E055_data <- H3K27me3_fc_E055[[4]]
# get subset data on chr22
gr.chr22 <- GRanges(seqnames = "chr22", ranges = IRanges(start =1, end = 51304566))
H3K27me3_fc.rel_E055 <- import(H3K27me3_fc_E055_data, which =gr.chr22, as = "Rle")
H3K27me3_fc.rel_E055.chr22 <- H3K27me3_fc.rel_E055$chr22
# identify region
region_E003 <- slice(H3K27me3_fc.rel.chr22, upper = 0.5)
region_E055 <- slice(H3K27me3_fc.rel_E055.chr22, lower = 2)
region_E003 <- as(region_E003, "IRanges")
region_E055 <- as(region_E055, "IRanges")
inter_region <- intersect(region_E003, region_E055)
sum(width(inter_region))
```

- 1869937

## Question 7: What is the average observed-to-expected ratio of CpG dinucleotides for CpG Islands on chromosome 22?
*CpG Islands are dense clusters of CpGs. The classic definition of a CpG Island compares the observed to the expected frequencies of CpG dinucleotides as well as the GC content.*
*Specifically, the observed CpG frequency is just the number of “CG” dinucleotides in a region.*
*The expected CpG frequency is defined as the frequency of C multiplied by the frequency of G divided by the length of the region.*

```{r}
# load hg19 CpG dataset
CpGIsland <- query(ah, c("hg19", "CpG Islands"))
CpGIsland_data <- CpGIsland[["AH5086"]]
# get data on chr22
CpGIsland_data.chr22 <- subset(CpGIsland_data, seqnames == "chr22")
CpGIsland_data.chr22.vi <- Views(Hsapiens, CpGIsland_data.chr22)
# calculate observed GC bases
region_length <- width(CpGIsland_data.chr22.vi)
observed_gcBase <- dinucleotideFrequency(CpGIsland_data.chr22.vi)[,7]/region_length
# calculate expected GC bases
freq_C <- letterFrequency(CpGIsland_data.chr22.vi, "C")
freq_G <- letterFrequency(CpGIsland_data.chr22.vi, "G")
expected_gcBase <- (freq_C/region_length)*(freq_G/region_length)
# calculate ration
mean(observed_gcBase/expected_gcBase)
```

- 0.8340929

## Question 8: How many TATA boxes are there on chr 22 of build hg19 of the human genome?
*A TATA box is a DNA element of the form “TATAAA”. Around 25% of genes should have a TATA box in their promoter. We will examine this statement.*

```{r}
# load hg19 build dataset
library(BSgenome.Hsapiens.UCSC.hg19)
TATA_boxes <- countPattern("TATAAA", Hsapiens$chr22) + countPattern("TATAAA", reverseComplement(Hsapiens$chr22))
TATA_boxes
```

- 27263

## Question 9: How many promoters of transcripts on chromosome 22 containing a coding sequence, contains a TATA box on the same strand as the transcript?

```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
gr <- GRanges(seqnames = "chr22", ranges = IRanges(start = 1, end = 52330658))
gr.trans.chr22 <- subsetByOverlaps(transcripts(txdb), gr, ignore.strand = TRUE)
proms <- promoters(gr.trans.chr22, upstream = 900, downstream = 100)
cdseq <- subsetByOverlaps(genes(txdb), gr, ignore.strand = TRUE)
proms_cds <- findOverlaps(proms, cdseq)
unique(queryHits(proms_cds))
count = 0
for (i in unique(queryHits(proms_cds))){
  proms_cds_vi <- Views(Hsapiens, proms[i])
  count = count + vcountPattern("TATAAA", DNAStringSet(proms_cds_vi))
}
count
```

- 266

## Question 10: How many bases on chr22 are part of more than one promoter of a coding sequence?
*It is possible for two promoters from different transcripts to overlap, in which case the regulatory features inside the overlap might affect both transcripts. This happens frequently in bacteria.*

```{r}
library(GenomicFeatures)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
seqlevels(txdb, force=TRUE) <- c("chr22")
gr <- GRanges(seqnames = "chr22", ranges = IRanges(start = 1, end = 52330658))
gr.trans.chr22 <- subsetByOverlaps(transcripts(txdb), gr, ignore.strand = TRUE)
length(gr.trans.chr22) 
gr.prom <- promoters(gr.trans.chr22, upstream = 900, downstream = 100)
tl.chr22 <- transcriptLengths(txdb, with.cds_len = TRUE) #rtn df
tl.chr22  <- tl.chr22[tl.chr22$cds_len > 0,]
trans.eval <- gr.prom[mcols(gr.prom)$tx_id %in% tl.chr22$tx_id]
sum(coverage(trans.eval) > 1)
```

- 306920