# 2. Sequences and Genomic Features

## Data Provided

- BAM file and its index
    - athal_wu_0_A.bam
    - athal_wu_0_A.bam.bai

- GTF file
    - athal_wu_0_A_annot.gtf

## File Setup

### Navigate to Project Folder

```bash
cd 2
```

### Extract Files

```bash
gunzip gencommand_proj2_data.tar.gz
tar xvc gencommand_proj2_data.tar
```

## Questions

### Question 1: How many alignments does the set contain?

```bash
samtools view athal_wu_0\_A.bam \| head
samtools view athal_wu_0\_A.bam \| wc -l

221372
```

### Question 2: How many alignments show the read's mate unmapped?

```bash
samtools view athal_wu_0_A.bam | cut -f7 | grep "*" | wc -l

65521
```

### Question 3: How many alignments contain a deletion (D)?

```bash
samtools view athal_wu_0_A.bam | cut -f6 | grep "D" | wc -l

2451
```

### Question 4: How many alignments show the read's mate mapped to the same chromosome?

```bash
samtools view athal_wu_0_A.bam | cut -f7 | grep "=" | wc -l

150913
```

### Question 5: How many alignments are spliced?

```bash
samtools view athal_wu_0_A.bam | cut -f6 | grep "N" | wc -l

0
```

### Question 6: How many alignments does the set contain?

```bash
samtools sort athal_wu_0_A.bam athal_wu_0_A.sorted 
samtools index athal_wu_0_A.sorted.bam
samtools view athal_wu_0_A.sorted.bam "Chr3:11777000-11794000" | head
samtools view athal_wu_0_A.sorted.bam "Chr3:11777000-11794000" | wc -l

7081
```

### Question 7: How many alignments show the read's mate unmapped?

```bash
samtools view athal_wu_0_A.sorted.bam "Chr3:11777000-11794000" | cut -f7 | grep "*" | wc -l

1983
```

### Question 8: How many alignments contain a deletion (D)?

```bash
samtools view athal_wu_0_A.sorted.bam"Chr3:11777000-11794000" | cut -f6 | grep "D" | wc -l

31
```

### Question 9: How many alignments show the read's mate mapped to the same chromosome?

```bash
samtools view athal_wu_0_A.sorted.bam "Chr3:11777000-11794000" | cut -f7 | grep "=" | wc -l

4670
```

### Question 10: How many alignments are spliced?

```bash
samtools view athal_wu_0_A.sorted.bam "Chr3:11777000-11794000" | cut -f6 | grep "N" | wc -l

0
```

### Question 11: How many sequences are in the genome file?

```bash
samtools view –H athal_wu_0_A.bam | grep –c “SN:”

7
```

### Question 12: What is the length of the first sequence in the genome file?

```bash
% samtools view –H athal_wu_0_A.bam | grep “SN:” | more

29923332
```

### Question 13: What alignment tool was used?

```bash
samtools view –H athal_wu_0_A.bam | grep “^@PG”

Stampy
```

### Question 14: What is the read identifier (name) for the first alignment?

### Question 15: What is the start position of this read's mate on the genome? Give this as 'chrom:pos' if the read was mapped, or '\*" if unmapped.

```bash
samtools view athal_wu_0_A.bam | head | cut -f1

GAII05_0002:1:113:7822:3886#0
```

### Question 16: How many overlaps (each overlap is reported on one line) are reported?

```bash
bedtools bamtobed -i athal_wu_0_A.bam > overlaps.bed
head overlaps.bed
bedtools intersect -wo -a athal_wu_0_A_annot.gtf -b overlaps.bed | head
bedtools intersect -wo -a athal_wu_0_A_annot.gtf -b overlaps.bed | wc -l

3101
```

### Question 17: How many of these are 10 bases or longer?

```bash
cut –f22 overlaps.bed | sort –nrk1 | grep –n “^9” | head -1

2899
```

### Question 18: How many alignments overlap the annotations?

```bash
cut –f1-12 overlaps.bed | sort –u | wc -l

3101
```

### Question 19: Conversely, how many exons have reads mapped to them?

```bash
cut –f13-21 overlaps.bed | sort –u | wc -l

1752
```

### Question 20: If you were to convert the transcript annotations in the file "athal_wu_0_A_annot.gtf" into BED format, how many BED records would be generated?

```bash
cut –f9 athal_wu_0_A.annot.gtf | cut –d ‘ ‘ –f1,2 | sort –u | wc -l

4
```