# 3. Alignment and Sequence Variation

## Data Provided

- FASTA file
    - wu_0.v7.fas

- FASTQ file
    - wu_0_A_wgs.fastq

## Instructions

For this project, it is recommended that you use the VMBox virtual environment (***Docker instead for MacOS with M1 or M2 chips) provided with the Course package and the tools therein. You may also use your own system and software, however make sure that appropriate versions are installed. The answers are compatible with the following versions of the software: samtools v.1.2, bowtie v.2.2.5 and bcftools v.1.2. 

As part of the effort to catalog genetic variation in the plant Arabidopsis thaliana, you re-sequenced the genome of one strain (‘wu_0_A’; genome file: ‘wu_0.v7.fas’), to determine genetic variants in this organism. The sequencing reads produced are in the file ‘wu_0_A_wgs.fastq’. Using the tools bowtie2, samtools and bcftools, develop a pipeline for variant calling in this genome. NOTE: Genome and re-sequencing data have been obtained and modified from those generated by the 1001 Genomes Project, accession ‘Wu_0_A’. 

### Questions 1 - 5:

Generate a bowtie2 index of the wu_0_A genome using bowtie2-build, with
the prefix 'wu_0'.

### Questions 6 - 14:

Run bowtie2 to align the reads to the genome, under two scenarios:
first, to report only full-length matches of the reads; and second, to
allow partial (local) matches. All other parameters are as set by
default.

**For the following set of questions (11 - 20), use the set of
full-length alignments calculated under scenario 1 only. Convert this
SAM file to BAM, then sort the resulting BAM file.**

### Questions 15 - 19:

Compile candidate sites of variation using SAMtools mpileup for further
evaluation with BCFtools. Provide the reference fasta genome and use the
option "-uv" to generate the output in uncompressed VCF format for easy
examination.

### Questions 20 - 24:

Call variants using 'BCFtools call' with the multiallelic-caller model.
For this, you will need to first re-run SAMtools mpileup with the BCF
output format option ('-g') to generate the set of candidate sites to be
evaluated by BCFtools. In the output to BCFtools, select to show only
the variant sites, in uncompressed VCF format for easy examination.

## File Setup

### Navigate to Project Folder

```bash
cd 3
```

### Extract Files

```bash
gunzip gencommand_proj3_data.tar.gz
tar xvc gencommand_proj3_data.tar
```

## Questions

### Question 1: How many sequences were in the genome?

```bash
ls -tl
head wu_0.v7.fas
cat wu_0.v7.fas | grep ">" | wc -l

7
```

### Question 2: What was the name of the third sequence in the genome file? Give the name only, without the ">" sign.

```bash
cat wu_0.v7.fas | grep ">" | head

\>Chr1
\>Chr2
\>Chr3
\>Chr4
\>Chr5
\>chloroplast
\>mitochondria

Chr3
```

### Question 3: What was the name of the last sequence in the genome file? Give the name only, without the ">" sign.

```bash
mitochondria
```

### Question 4: How many index files did the operation create?

```bash
mkdir wu_0
bowtie2-build wu_0.v7.fas wu_0/wu_0
ls wu_0 | wc -l

6
```

### Question 5: What is the 3-character extension for the index files created?

```bash
mkdir wu_0
bowtie2-build wu_0.v7.fas wu_0/wu_0
ls wu_0

.bt2
```

### Question 6: How many reads were in the original fastq file?

```bash
head wu_0_A_wgs.fastq
cat wu_0_A_wgs.fastq | wc -l
(note: the reads are # of lines / 4)

147354 (after dividing by 4)
```

### Question 7: How many matches (alignments) were reported for the original (full-match) setting? Exclude lines in the file containing unmapped reads. 

```bash
bowtie2 -p 4 -x wu_0/wu_0 wu_0_A_wgs.fastq -S wu_0.bt2.sam

147354 reads; of these:
147354 (100.00%) were unpaired; of these:
9635 (6.54%) aligned 0 times
93780 (63.64%) aligned exactly 1 time
43939 (29.82%) aligned >1 times
93.46% overall alignment rate

bowtie2 -p 4 --local -x wu_0/wu_0 wu_0_A_wgs.fastq -S wu_0.bt2.local.sam

147354 reads; of these:
147354 (100.00%) were unpaired; of these:
6310 (4.28%) aligned 0 times
84939 (57.64%) aligned exactly 1 time
56105 (38.07%) aligned >1 times
95.72% overall alignment rate

137719 (93780 + 43939)
```

### Question 8: How many matches (alignments) were reported with the local-match setting? Exclude lines in the file containing unmapped reads. 

```bash
141044 (84939 + 56105)
```

### Question 9: How many reads were mapped in the scenario in Question 7?

```bash
137719
```

### Question 10: How many reads were mapped in the scenario in Question 8?

```bash
141044
```




### Question 11: How many reads had multiple matches in the scenario in Question 7? You can find this in the bowtie2 summary; note that by default bowtie2 only reports the best match for each read.

```bash
43939
```

### Question 12: How many reads had multiple matches in the scenario in Question 8? Use the format above. You can find this in the bowtie2 summary; note that by default bowtie2 only reports the best match for each read.

```bash
56105
```

### Question 13: How many alignments contained insertions and/or deletions, in the scenario in Question 7?

```bash
cut –f6 wu_0.bt2.sam | grep –c “[I,D]”

2782
```

### Question 14: How many alignments contained insertions and/or deletions, in the scenario in Question 8? 

```bash
cut –f6 wu_0.bt2.local.sam | grep –c “[I,D]”

2614
```

### Question 15: How many entries were reported for Chr3?

```bash
samtools view -bT wu_0.v7.fas wu_0.bt2.sam > wu_0.bt2.bam
samtools sort wu_0.bt2.bam wu_0.bt2.sorted.bam
samtools index wu_0.bt2.sorted.bam.bam
samtools mpileup -f wu_0.v7.fas wu_0.bt2.sorted.bam.bam > wu_0.mpileup
samtools mpileup -uv -f wu_0.v7.fas wu_0.bt2.sorted.bam.bam > wu_0.vcf
more wu_0.vcf
cat wu_0.vcf | cut -f1 | grep "Chr3" | head

/##contig=<ID=Chr3,length=23042017>
Chr3
Chr3
Chr3
Chr3
Chr3
Chr3
Chr3
Chr3
Chr3

cat wu_0.vcf | cut -f1 | grep "Chr3" | wc -l

360295 (360296 - 1 for header)
```

### Question 16: How many entries have 'A' as the corresponding genome letter?

```bash
cat out.full.mpileup.vcf | grep –v “^#” | cut –f4 | grep –P “^A$”

1150985
```

### Question 17: How many entries have exactly 20 supporting reads (read depth)?

```bash
cat wu_0.vcf | cut -f8 | grep -c "DP=20"

1816
```

### Question 18: How many entries represent indels?

```bash
cat wu_0.vcf | cut -f8 | grep "INDEL" | head

1972 (1973 - 1 for header)
```

### Question 19: How many entries are reported for position 175672 on Chr1?

```bash
cat wu_0.vcf | cut -f1,2 | grep "Chr1" | awk '$2 == "175672"' | head

Chr1 175672
Chr1 175672

cat wu_0.vcf | cut -f1,2 | grep "Chr1" | awk '$2 == "175672"' | wc -l

2
```

### Question 20: How many variants are called on Chr3?

```bash
samtools mpileup -g -f wu_0.v7.fas wu_0.bt2.sorted.bam.bam > wu_0.bcf
bcftools call -v -m -O z -o wu_0.vcf.gz wu_0.bcf
zcat wu_0.vcf.gz | cut -f1 | grep "Chr3" | head

/##contig=<ID=Chr3,length=23042017>
Chr3
Chr3
Chr3
Chr3
Chr3
Chr3
Chr3
Chr3
Chr3

zcat wu_0.vcf.gz | cut -f1 | grep "Chr3" | wc -l

398 (399 - 1 for header)
```

### Question 21: How many variants represent an A->T SNP? If useful, you can use 'grep --P' to allow tabular spaces in the search term.

```bash
zcat wu_0.vcf.gz | cut -f4,5 | awk '$1 == "A" && $2 == "T"' | head

A T
A T
A T
A T
A T
A T
A T
A T
A T
A T

zcat wu_0.vcf.gz | cut -f4,5 | awk '$1 == "A" && $2 == "T"' | wc -l

392
```

### Question 22: How many entries are indels?

```bash
zcat wu_0.vcf.gz | cut -f8 | grep -c "INDEL" | head

320 (321 - 1 for header)
```

### Question 23: How many entries have precisely 20 supporting reads (read depth)?

```bash
zcat wu_0.vcf.gz | cut -f8 | grep "DP=20" | head

DP=20;VDB=0.0587288;SGB=-0.556411;RPB=0.639909;MQB=0.931063;BQB=0.972484;MQ0F=0;ICB=1;HOB=0.5;AC=1;AN=2;DP4=16,0,4,0;MQ=39

DP=20;VDB=0.255089;SGB=-0.676189;RPB=0.97436;MQB=0.499893;BQB=0.850154;MQ0F=0.05;ICB=1;HOB=0.5;AC=1;AN=2;DP4=9,0,11,0;MQ=13

zcat wu_0.vcf.gz | cut -f8 | grep "DP=20" | wc -l

2
```

### Question 24: What type of variant (i.e., SNP or INDEL) is called at position 11937923 on Chr3?

```bash
zcat wu_0.vcf.gz | grep "Chr3" | awk '$2 == "11937923"'

Chr3 11937923 . G A 13.73 .
DP=20;VDB=0.0587288;SGB=-0.556411;RPB=0.639909;MQB=0.931063;BQB=0.972484;MQ0F=0;ICB=1;HOB=0.5;AC=1;AN=2;DP4=16,0,4,0;MQ=39
GT:PL 0/1:48,0,137

SNP (G --> A)
```
